JB/gallery screen

jmbidev opened the pull request:
![Screenshot_1582814573](https://user-images.githubusercontent.com/48542614/75460576-61288780-5960-11ea-8ef3-31ec113f66bc.png)

jmbidev committed:
Gallery screen implementation

jmbidev committed:
Add permissions management

jmbidev committed:
Fix catch exception on not selected image

MPucheta commented on src/screens/MainApp/DiagnoseRequest/screens/Gallery/components/ListItem/styles.js:
```diff
@@ -0,0 +1,40 @@
+  },
+  badgeText: {
+    fontWeight: 'bold',
+    textAlign: 'center'
```
I'm concerned this `fontWeight: 'bold'` reverts the font to the default one (React Native problem regarding custom fonts), It may not be noticeable as the custom font is similar to the default one but it's definitely something to check for. 

This `bold` for `Red Hat Display` font only seems to appear in this case though. It is not reflected in the styleguide.

Feel free to either use the `Raleway bold` (`SubHeader`) or add a `Red Hat Display Bold`  with a new text component

MPucheta commented on src/screens/MainApp/DiagnoseRequest/screens/Gallery/components/ListItem/styles.js:
```diff
@@ -0,0 +1,40 @@
+    marginBottom: scale(1),
+    marginRight: scale(3),
+    alignContent: 'center',
+    backgroundColor: '#5ECE84',
```
This is `theme.colors.primary`

MPucheta commented on src/screens/MainApp/DiagnoseRequest/screens/Gallery/components/ListItem/styles.js:
```diff
@@ -0,0 +1,40 @@
+  imageButton: {
+    padding: scale(4),
+    margin: scale(4),
+    backgroundColor: '#5ECE84',
```
Same here, please change this to `theme.colors.primary`

MPucheta commented on src/screens/MainApp/DiagnoseRequest/screens/Gallery/index.js:
```diff
@@ -0,0 +1,197 @@
+import { FlatList, View, Alert, PermissionsAndroid } from 'react-native'
+import { SvgXml } from 'react-native-svg'
+import Background from '~/components/Background'
+import { PrimaryButton, Description, Button, Body } from '~/components'
```
I would change this to more specific imports

`import { PrimaryButton, Button } from '~/components/buttons'`
`import { Description, Body } from '~/components/texts'`

MPucheta commented on src/screens/MainApp/DiagnoseRequest/screens/Gallery/index.js:
```diff
@@ -0,0 +1,197 @@
+
+const requestWritePermissions = async () => {
+  return requestPermissions(PermissionsAndroid.PERMISSIONS.WRITE_EXTERNAL_STORAGE)
+}
```
Please consider moving this to a separate folder like `permissions` or `permissionService`

MPucheta commented on src/screens/MainApp/DiagnoseRequest/screens/Gallery/index.js:
```diff
@@ -0,0 +1,197 @@
+      const granted = await requestWritePermissions()
+
+      if (granted !== PermissionsAndroid.RESULTS.GRANTED) {
+        alertPermissionsError()
```
Add a blank before the return

MPucheta commented on src/screens/MainApp/DiagnoseRequest/screens/Gallery/index.js:
```diff
@@ -0,0 +1,197 @@
+      const newImages = [...images]
+      newImages.splice(1, 0, ...params.addedImages)
+      setImages(newImages)
+    }
```
Add a blank as separation (I consider this two different logical groupings)

MPucheta commented on src/screens/MainApp/DiagnoseRequest/screens/Gallery/index.js:
```diff
@@ -0,0 +1,197 @@
+                  <SvgXml xml={CameraLogo} />
+                  <Body primary>Sacar foto</Body>
+                </View>
+              </Button>
```
Not necessary but you could probably extract this to a `TakePhotoButton` component.

Another possible name: `TakePhotoCTA`

MPucheta commented on src/screens/MainApp/DiagnoseRequest/screens/Gallery/index.js:
```diff
@@ -0,0 +1,197 @@
+                </View>
+              </Button>
+            )
+          }
```
Add a blank line before `return`

MPucheta commented on src/screens/MainApp/DiagnoseRequest/screens/Gallery/index.js:
```diff
@@ -0,0 +1,197 @@
+        >
+          <Description white>Continuar</Description>
+        </PrimaryButton>
+      )}
```
Likewise you can move this to a separate `goToReviewButton`.

This case is more nitpicky though, it reads fine and the possibility of it changing is low.

MPucheta commented on src/screens/MainApp/DiagnoseRequest/screens/Gallery/styles.js:
```diff
@@ -0,0 +1,33 @@
+import { StyleSheet } from 'react-native'
+import { scale } from 'react-native-size-matters/extend'
+import { verticalScale } from 'react-native-size-matters'
```
Use only `react-native-size-matters/extend`

MPucheta commented on src/screens/MainApp/DiagnoseRequest/screens/Gallery/styles.js:
```diff
@@ -0,0 +1,33 @@
+    margin: scale(4),
+    borderRadius: scale(6),
+    backgroundColor: '#5ECE841A',
+    borderColor: '#5ECE84',
```
This is `theme.colors.primary`

MPucheta commented on src/screens/MainApp/DiagnoseRequest/screens/Gallery/styles.js:
```diff
@@ -0,0 +1,33 @@
+  box: {
+    margin: scale(4),
+    borderRadius: scale(6),
+    backgroundColor: '#5ECE841A',
```
Wow i did not know this was a thing (`#rrggbbaa`), consider creating a `theme.colors.transparentPrimary` or `theme.colors.primaryAlpha`

MPucheta commented on src/screens/MainApp/DiagnoseRequest/screens/ImageReview/index.js:
```diff
@@ -122,18 +109,25 @@ const ImageVisualization = ({ navigation }) => {
+        getMorePhotos()
+      }
+    )
+    return () => {
+      handler.remove()
+    }
```
Just a thought, should we able to just return `handler.remove`?

nahudalla replied:
If `remove` is a method, you'd be passing the function only and it'd be called without the `this` keyword attached to `handler`.
```javascript
const remove = handler.remove
const boundRemove = handler.remove.bind(remove)

remove() // can fail because of "this" not being bound
handler.remove() // will succeed
boundRemove() // will succeed
```

MPucheta reviewed (changes requested):
Excellent job, i really like this addition!

I left some comments,  many are calls to refactorings, general organization and minor fixes.

There is something to address though, taking a new photo while on gallery, which adds a new image to the state, causes all other images to flicker with a primary color background. This is somewhat expected but there should be a workaround. I suspect there may be a problem with the flatlist key extractor and the concrete `ListItem` component, check in `devtools` if the key is properly attached to the component too.

jmbidev committed:
Change font to raleway-bold, use theme colors in styles, move permissions functionality to permissionsService and fix imports.

MPucheta committed:
Refactor code and fix new photo flickering in gallery

Flickering was caused by a very specific problem regarding CellRenderer keys. Only solution was to use flexWrap property for contentContainerStyle in FlatList, which causes a REALLY annoying warning in __DEV__. I added the necessary logic to turn that warning off.

There still is a problem, in __PROD__ the screen works wonderfully, but in __DEV__ it's kind of slow. Further research will be done.
